numpy/_core/src/multiarray/stringdtype/casts.c

```diff
@@ -0,0 +1,1510 @@
+
+    return -1;
+}
+
+static PyType_Slot s2s_slots[] = {
+        {NPY_METH_resolve_descriptors, &string_to_string_resolve_descriptors},
+        {NPY_METH_strided_loop, &string_to_string},
+        {NPY_METH_unaligned_strided_loop, &string_to_string},
+        {0, NULL}};
+
+static char *s2s_name = "cast_StringDType_to_StringDType";
+
+// unicode to string
+
+// Find the number of bytes, *utf8_bytes*, needed to store the string
+// represented by *codepoints* in UTF-8. The array of *codepoints* is
+// *max_length* long, but may be padded with null codepoints. *num_codepoints*
+// is the number of codepoints that are not trailing null codepoints. Returns
+// 0 on success and -1 when an invalid code point is found.
+static int
+utf8_size(const Py_UCS4 *codepoints, long max_length, size_t *num_codepoints,
+          size_t *utf8_bytes)
+{
+    size_t ucs4len = max_length;
+
+    while (ucs4len > 0 && codepoints[ucs4len - 1] == 0) {
+        ucs4len--;
+    }
+    // ucs4len is now the number of codepoints that aren't trailing nulls.
+
+    size_t num_bytes = 0;
+
+    for (size_t i = 0; i < ucs4len; i++) {
+        Py_UCS4 code = codepoints[i];
+
+        if (code <= 0x7F) {
+            num_bytes += 1;
+        }
+        else if (code <= 0x07FF) {
+            num_bytes += 2;
+        }
+        else if (code <= 0xFFFF) {
+            if ((code >= 0xD800) && (code <= 0xDFFF)) {
+                // surrogates are invalid UCS4 code points
+                return -1;
+            }
+            num_bytes += 3;
+        }
+        else if (code <= 0x10FFFF) {
+            num_bytes += 4;
+        }
+        else {
+            // codepoint is outside the valid unicode range
+            return -1;
+        }
+    }
+
+    *num_codepoints = ucs4len;
+    *utf8_bytes = num_bytes;
+
+    return 0;
+}
+
+// Converts UCS4 code point *code* to 4-byte character array *c*. Assumes *c*
+// is a zero-filled 4 byte array and *code* is a valid codepoint and does not
+// do any error checking! Returns the number of bytes in the UTF-8 character.
+static size_t
+ucs4_code_to_utf8_char(const Py_UCS4 code, char *c)
+{
+    if (code <= 0x7F) {
+        // 0zzzzzzz -> 0zzzzzzz
+        c[0] = (char)code;
+        return 1;
+    }
+    else if (code <= 0x07FF) {
+        // 00000yyy yyzzzzzz -> 110yyyyy 10zzzzzz
+        c[0] = (0xC0 | (code >> 6));
+        c[1] = (0x80 | (code & 0x3F));
+        return 2;
+    }
+    else if (code <= 0xFFFF) {
+        // xxxxyyyy yyzzzzzz -> 110yyyyy 10zzzzzz
+        c[0] = (0xe0 | (code >> 12));
+        c[1] = (0x80 | ((code >> 6) & 0x3f));
+        c[2] = (0x80 | (code & 0x3f));
+        return 3;
+    }
+    else {
+        // 00wwwxx xxxxyyyy yyzzzzzz -> 11110www 10xxxxxx 10yyyyyy 10zzzzzz
+        c[0] = (0xf0 | (code >> 18));
+        c[1] = (0x80 | ((code >> 12) & 0x3f));
+        c[2] = (0x80 | ((code >> 6) & 0x3f));
+        c[3] = (0x80 | (code & 0x3f));
+        return 4;
+    }
+}
```

Comment:
Happy to keep these here, they do also look pretty good as their own file with utf8 utilities.